### ISSUE TITLE
FR: Add info window or some progress indicator when opening folder with many files

### ISSUE THREAD

**Issue by sonix-github** [koreader/koreader#9793]
Hello KOReader team,

it would be nice, that during opening some folder with many files (>~500), will pop-up some info window or progress indicator to the user so he/she get feedback that the program is doing something more time consuming.

What do you think?

Thanks for any constructive feedback.

**Does your feature request involve difficulty completing a task? Please describe.**
I think it can be used the same idea which is used when opening an e-book.

**Describe the solution you'd like**
In file manager after tap to open folder wait 1-2 seconds and then display a info message about opening/reading the files or show some progress indicator during it.

**Comment by NiLuJe** [comment]
I'm not entirely sure there's anything we can do, as it's entirely IO bound, so we're literally stuck waiting for the disk ;).

The CRe stuff works because CRe itself sends us a ping as it progresses along. Here, not so much, we're stuck waiting for syscalls to return.

**Comment by NiLuJe** [comment]
I'm not familiar with the codepaths in question, but you don't really have an idea of how deep a directory walk is going to be going in, you get thrown stuff as it comes up, so at *best*, I could imagine some sort of spinner, but a true linear progress bar, probably not.

(Extremely low priority in my book FWIW, hence the label)

**Comment by NiLuJe** [comment]
I can't quite think of how desktop filemanagers handle this sort of situation. For pretty much most of the same reasons, I'm guessing they... don't? ;o).

RFC ;).

**Comment by Frenzie** [comment]
Due to the paged nature of things you might be able to delay getting most of the file attributes a fair bit, but that'd be an awful lot of added complexity for little benefit. 

You left out the device btw.

**Comment by sonix-github** [comment]
Thanks @NiluJe for clarification. In that case it would be sufficient to show small info message like "Opening directory. Please wait..." or so just after tap directory from file manager.

**Comment by sonix-github** [comment]
> You left out the device btw.

Ohh. Sorry. I am using Pocketbook PB626

**Comment by hius07** [comment]
What is the display mode? If it is not "classic" the books metadata is fetched.

**Comment by Frenzie** [comment]
The metadata only happens on a page per page basis.

But the attributes are all preloaded immediately unless I'm mistaken. (Which is much faster, up to a point.)

**Comment by sonix-github** [comment]
> What is the display mode?

Currently I have "Detailed lest with cover images and metadata"

**Comment by offset-torque** [comment]
> **NiLuJe** I'm not familiar with the codepaths in question, but you don't really have an idea of how deep a directory walk is going to be going in, you get thrown stuff as it comes up, so at best, I could imagine some sort of spinner, but a true linear progress bar, probably not.

@NiLuJe
We don't know how long it will take but we know how many items (books, directories) is already processed, right ?

The popup can just say **`X files processed`** and increment that X as it walks through the directory.

**Benefits:**
- It will signalize that KOReader is not frozen, still processing
- **"Increasing counter"** gives the user something to look at while waiting to finish
- No extra time calculation/progress bar involved, easier to implement

Update can even be every second (instead of real-time) to prevent extra CPU usage (if a restriction like this is necessary, I don't know).

**Comment by NiLuJe** [comment]
Nothing is being "processed" though, we're *literally* waiting on a syscall (possibly a *single* one, making any and all progress report unfeasible).

Actual porcessing only happens for metadata, and that's already done on a page-by-page basis, and backgrounded, so, not an issue in practice.

**Comment by NiLuJe** [comment]
It's all castles in the sky until someone actually gets us data of what's actually stalling and where, though.

**Comment by offset-torque** [comment]
I see. In that case probably best option is a simple popup like "**Processing, please wait...**"

**Comment by NiLuJe** [comment]
That would just be *extremely* irritating noise in 99.999% of the cases :/.

**Comment by NiLuJe** [comment]
(i.e., this should *very* much be a "do it right, or don't do it at all" affair).

**Comment by NiLuJe** [comment]
The issue is that if it's a single syscall that hangs for a while, we won't ever know about it until it returns (i.e., *after* the stall).

Which is why I'm pessimistic about this. I'd need way more data to be able to say "okay, we might be able to do something about it".

We might. depends on exactly what happens. But I really don't think there's much we can do ;).

**Comment by Frenzie** [comment]
@NiLuJe I'm not sure if it's due to the context of the discussion but iirc I'm not sure if you're describing it correctly, unless someone changed it in the meantime.

There's `lfs.dir()` (or whatever) which nothing can be done about.

But with the results we do `lfs.attributes()` one by one.

For certain sorting methods, I'm sure you can relatively easily do that in batches of 1 or 2 pages ahead, adding the rest more dynamically. The other stuff like size and modification doesn't matter for sort by filename.

**Comment by sonix-github** [comment]
Just an idea. Is it possible to show the "processing" message only if the directory processing does not finish within ~2 seconds or so? I mean start a timer just before directory processing and after preset time it shows the "processing" message or if the file processing finishes sooner, it will stop the timer, so therefore no message will be shown.

**Comment by Frenzie** [comment]
Not really.

As an aside, here's an awk script to create files on busybox, yanked from https://askubuntu.com/a/689248.

```sh
# "normal", 999 files, takes about 2-3 seconds to open on my H2O
awk 'BEGIN{for(i=ARGV[2];i<=ARGV[3];i++){fd=sprintf("%s%04d.txt",ARGV[1],i); printf "" > fd;close(fd)}}' bslp 1 999

# absurd, 9999 files, even severely slows down just showing the parent dir
# opening the directory is never-ending rather than simply 20-30 seconds
# not sure what that means
awk 'BEGIN{for(i=ARGV[2];i<=ARGV[3];i++){fd=sprintf("%s%04d.txt",ARGV[1],i); printf "" > fd;close(fd)}}' bslp 1 999
```

**Comment by poire-z** [comment]
(I think this would not need more thinking - users should learn the hard way it's a bad idea to put 500+ files in a folder, and change their habits and get tidy and decent human beings :)

But for those interested, it looks like we don't have a single syscall: we use lfs.dir(), which returns an iterator - and we just use it in Lua land to iterate all this directory entries, and grab some info about each file.
So, it might be possible to hook a timer in there and check if it some time has elapsed when we are still iterating, to call some callback that would display some indicator.
Feels a bit ugly to plug some UI stuff in such a low level function though.

https://github.com/koreader/koreader/blob/a23c30f862aea0cd553fa7491edfc68224e931be/frontend/ui/widget/filechooser.lua#L103-L111

**Comment by Frenzie** [comment]
> (I think this would not need more thinking - users should learn the hard way it's a bad idea to put 500+ files in a folder, and change their habits and get tidy and decent human beings :)

Agreed, I don't know how you'd find anything. Performance wise a thousand seems perfectly fine though, if a tad slower than desirable.

**Comment by NiLuJe** [comment]
@poire-z: Yeah, my assumptions yesterday was that it was the initial *listing* that would be slow, not the individual stat calls.

But, given that we do compute the amount of items in a folder... somewhere, possibly not?

Anyway, definitely agree that:

* Putting UI code in there would be ugly as sin
* > 500 entries in a folder on FAT32 is a terrible idea (the Kindle system will actually warn about it and tell you to gtfo).

**Comment by NiLuJe** [comment]
IIRC how readdir works, the first readdir calls does *significantly* more work than all the others, this is what I was thinking of ;).

**Comment by hius07** [comment]
> * the Kindle system will actually warn about

The message is:

**Folder is almost full**
One of the subfolders under Documents folder has exceeded 2000 items.
To continue adding more items to your device, we recommend creating
a new subfolder under Downloads.

**Comment by Frenzie** [comment]
> Over 500 entries in a single folder on FAT32 is a terrible idea (the Kindle system will actually warn about it and tell you to get your act together and stop shooting yourself in the foot).

As I said, that takes a second on Kobo. Don't generalize Kindle sucking. ;-)
